wifiphisher/common/extensions.py

```diff
@@ -0,0 +1,194 @@
+            obj = Class(shared_data)
+            self._extensions.append(obj)
+
+    def start_extensions(self):
+        """
+        Starts the two main daemons of EM:
+
+        1) Daemon that listens to every packet and 
+        forwards it to each extension for further processing.
+        2) Daemon that receives special-crafted packets
+        from extensions and broadcasts them in the air.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        # One daemon is listening for packets...
+        listen_thread = threading.Thread(target=self._listen)
+        listen_thread.start()
+        # ...another daemon is sending packets
+        send_thread = threading.Thread(
+            target=self._send)
+        send_thread.start()
+
+    def on_exit(self):
+        """
+        Stops both daemons of EM on exit.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        self._should_continue = False
+
+    def get_output(self):
+        """
+        Gets the output of each extensions.
+        Merges them in a list and returns it.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :return: None
+        :rtype: None
+        """
+
+        output = []
+        for m in self._extensions:
+            m_output = m.send_output()
+            if m_output and len(m_output) > 0:
+                output += m_output
+        return output
+
+    def _process_packet(self, pkt):
+        """
+        Pass each captured packet to each module.
+        Gets the packets to send.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param pkt: A Scapy packet object
+        :type pkt: Scapy Packet
+        :return: None
+        :rtype: None
+        """
+
+        for m in self._extensions:
+            received_packets = m.get_packet(pkt)
+            if received_packets and len(received_packets) > 0:
+                self._packets_to_send += received_packets
+
+    def _listen(self):
+        """
+        Listening thread. Listens for packets and forwards them
+        to _process_packet.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager 
+        :return: None
+        :rtype: None
+        """
+
+        # continue to find clients until told otherwise
+        while self._should_continue:
+            dot11.sniff(iface=self._interface, prn=self._process_packet,
+                        count=1, store=0)
+
+    def _send(self):
+        """
+        Sending thread. Continously broadcasting packets
+        crafted by extensions.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager 
+        :return: None
+        :rtype: None
+        """
+
+        while self._should_continue:
```

Comment:
Yes. We want to keep checking if there are any packets to send (first loop). 

If there are, we want to keep broadcasting packets (second loop). However in the second loop we don't want to include the first "if statement" for performance reasons.